File: ArchitectureHW5/pinSharper.cpp
#include <iostream>
#include "pinSharper.h"

// Constructor.
pinSharper::pinSharper(pinSharpChecker *sharp_check, std::atomic<int> *pins_left) {
    sharp_check_ = sharp_check;
    pins_left_ = pins_left;
}

// Thread run.
void pinSharper::run() {
    while (true) {
        lock_.lock();
        if (*pins_left_ < 1) {
            lock_.unlock();
            return;
        }

        // Checking if program still needs to work.
        if (*pins_left_ < 1) {
            lock_.unlock();
            return;
        }

        // If no new pins came for now.
        if ((pins_.empty()) || (pins_.back() == nullptr)) {
            lock_.unlock();
            continue;
        }

        // Checking if program still needs to work.
        if (*pins_left_ < 1) {
            lock_.unlock();
            return;
        }

        // Sharpening the pin.
        pins_.back()->sharpness = pins_.back()->sharpness * 2;
        if (pins_.back()->sharpness > 1) {
            // Maximum pin sharpness is 1.
            pins_.back()->sharpness = 1;
        }

//        std::cout << "Grinder man sharped a pin, now it has curvature " +
//                  std::to_string(pins_.back()->curvature) + " and sharpness " +
//                  std::to_string(pins_.back()->sharpness) << '\n';

        lock_.unlock();

        // Sending further.
        sendPin(sharp_check_);
    }
}

// Getting a new pin.
void pinSharper::receivePin(pin *pn) {
    lock_.lock();

    // Checking if program still needs to work.
    if (*pins_left_ < 1) {
        lock_.unlock();
        return;
    }

    // Adding a new pin to our current collection.
    pins_.push_back(pn);
    lock_.unlock();
}

// Giving away a sharpened pin.
void pinSharper::sendPin(pinSharpChecker *sharp_check) {
    lock_.lock();

    // Checking if program still needs to work.
    if (*pins_left_ < 1) {
        lock_.unlock();
        return;
    }

    // Sending the pin further.
//    std::cout << "Grinder man gave sharp checker a pin with curvature " +
//              std::to_string(pins_.back()->curvature) + " and sharpness " +
//              std::to_string(pins_.back()->sharpness) + '\n';

    sharp_check->receivePin(pins_.back());

    pins_.pop_back();
    lock_.unlock();
}

File: GoMultiithreading/factory/pinSharper.go
package factory

import (
	"go.uber.org/atomic"
	"math"
	"sync"
)

type PinSharper struct {
	// Pins which are currently being sharpened.
	pins []*Pin
	// The sharp checker whom grinder man is going to give pins.
	pinSharpChecker *PinSharpChecker
	// Thread lock.
	lock sync.Mutex
	// Number of total pins, which are still in process of approving / sharpening.
	pinsLeft *atomic.Uint32
}

// Constructor.
func NewSharp(checker *PinSharpChecker, pinsLeft *atomic.Uint32) *PinSharper {
	sharp := new(PinSharper)
	sharp.pinSharpChecker = checker
	sharp.pinsLeft = pinsLeft
	return sharp
}

// Thread run.
func (sharp *PinSharper) Run(wg *sync.WaitGroup) {
	// Worker has done his job.
	defer wg.Done()
	for sharp.pinsLeft.Load() > 0 {
		sharp.lock.Lock()

		// If worker has no pins for now.
		if len(sharp.pins) == 0 {
			sharp.lock.Unlock()
			continue
		}

		// Sharpening the pin.
		sharp.pins[len(sharp.pins)-1].sharpness = math.Min(sharp.pins[len(sharp.pins)-1].sharpness*2, 1)
		// fmt.Printf("Grinder man sharped a pin, now it has curvature %f and sharpness %f\n",
		// 	 sharp.pins[len(sharp.pins)-1].curvature,
		//	 sharp.pins[len(sharp.pins)-1].sharpness)

		sharp.lock.Unlock()
		sharp.sendPin(sharp.pinSharpChecker)
	}
}

// Getting a new pin.
func (sharp *PinSharper) receivePin(pin *Pin) {
	sharp.lock.Lock()
	// Adding a new pin to our current collection.
	sharp.pins = append(sharp.pins, pin)
	sharp.lock.Unlock()
}

// Giving away an approved pin.
func (sharp *PinSharper) sendPin(checker *PinSharpChecker) {
	sharp.lock.Lock()
	// fmt.Printf("Grinder man gave sharp checker a pin with curvature %f and sharpness %f\n",
	// 	 sharp.pins[len(sharp.pins)-1].curvature,
	// 	 sharp.pins[len(sharp.pins)-1].sharpness)
	// Sending the pin.
	checker.receivePin(sharp.pins[len(sharp.pins)-1])
	// The pin was sent.
	sharp.pins = sharp.pins[:len(sharp.pins)-1]
	sharp.lock.Unlock()
}
